pasta_eln/database.py

```diff
@@ -43,11 +44,16 @@ def __init__(self, user:str, password:str, databaseName:str, configuration:dict[
       self.db.create_document(self.ontology)
       self.initDocTypeViews( configuration['tableColumnsMax'] )
       self.initGeneralViews()
-    self.ontology = self.db['-ontology-']
-    if '-version' not in self.ontology or self.ontology['-version']!=2:
+    self.ontology = dict(self.db['-ontology-'])
+    if '-version' in self.ontology and self.ontology['-version']==2:
+      logging.info('Convert ontology version 2 to 3')
+      ontologyV2_to_V3(self.ontology)
```

Comment:
Not sure if this function call is doing something to the ontology document, actually when I tested the editor tool is not given a document, instead it get's a dict and save call is throwing exception, could you check this also?

![image](https://github.com/PASTA-ELN/pasta-eln/assets/141253011/7c179e3c-1c9a-468b-b9ed-4bfaafcde60e)

Reply:
But the ontology_configuration is designed presently in a way that the its expects the ontology data in a document format  and every edits done on UI will be reflected directly within the document dictionaries and finally when the user click on save button it simply calls the save functionality of the document to persists the changes

But now your recommendation is to keep a local copy of dict and do the edits on the local copy of dict and later before the save is called, copy all the changes from local dict to the document and save it? 

But what is the benefit of keeping local copy rather than doing the edits directly on the document? This is an extra effort to reflect the changes from local copy to the db document, Also then the editor needs to do the database calls and load the document directly? I think we should discuss this first since this is a major design change for ontology editor, see the pros and cons and then we must take decision there-after..

Reply:
I fully agree that it is extra effort to do things in a local dict and without benefit. I also would not like to do it. I just have the experience that reading-updating-reading documents with this library causes issues because library is not amazing. Therefore, I always do the detour. 

BTW, the editor does not need to re-read it when loading, the latest ontology-dictionary is in backend.db.ontology. Just for saving it needs to do couchdb access.
But if you want to stay in document world you can. Just read the ontology from the db on load, keep the existing code, save it to DB afterwards. If you want, you can try to read-update-update to check if the document database does as expected.
It is your choice.

Reply:
I need to know more of the problems which you have faced while using the document directly, also I do not have so much experience in using the library, I need to understand a bit more regarding the approach to be used for saving the document, shall we discuss this tomorrow during the call? I think, this would be helpful...

Reply:
As per the discussion, this issue will be addressed separately as part of [Save fails in ontology editor #42
](https://github.com/PASTA-ELN/pasta-eln/issues/42)